ai/core/src/unitycatalog/ai/core/utils/callable_utils.py

```diff
@@ -769,3 +772,216 @@ def check_docstring_signature_consistency(
             UserWarning,
             stacklevel=2,
         )
+
+
+def _recursively_parse_type(type_info: Any) -> str:
+    """
+    Recursively parse a SQL type definition (provided either as a dict or a string)
+    into a Python type annotation string.
+
+    Args:
+        type_info: A dict containing the type information, or a simple string.
+
+    Returns:
+        A string representing the Python type.
+    """
+    if not isinstance(type_info, dict):
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(str(type_info).upper(), None)
+        # NB: OSS Unity Catalog doesn't retain the inner subtypes for collections
+        # so the logic below for handling arrays and maps is unnecessary.
+        # We can only retrieve the outer type and will short circuit early here.
+        if isinstance(mapped, tuple):
+            mapped = mapped[0]
+        return mapped.__name__ if mapped else "Any"
+
+    t = type_info.get("type")
+    if not t:
+        return "Any"
+    t_lower = str(t).lower()
+
+    if t_lower == "array":
+        element = type_info.get("elementType")
+        return f"list[{get_mapped_value(element)}]"
+    elif t_lower == "map":
+        key = type_info.get("keyType")
+        value = type_info.get("valueType")
+        return f"dict[{get_mapped_value(key)}, {get_mapped_value(value)}]"
+    elif t_lower == "struct":
+        return "dict"
+    else:
+        mapped = SQL_TYPE_TO_PYTHON_TYPE_MAPPING.get(str(t).upper(), None)
+        return mapped.__name__ if mapped else "Any"
+
+
+def get_mapped_value(x: Any) -> str:
```

Comment:
nit: should this be a public function? The same question applies to methods other than dynamically_construct_python_function

Reply:
Updated all of them to be private - there's no reason to expose these as public utils